vision_transformer/main.py

```diff
@@ -0,0 +1,244 @@
+        self.dropout = args.dropout
+
+        self.embedding = InputEmbedding(args)
+        # Encoder Stack
+        self.encoders = nn.ModuleList([EncoderBlock(args) for _ in range(self.num_encoders)])
+        self.MLPHead = nn.Sequential(
+            nn.LayerNorm(self.latent_size),
+            nn.Linear(self.latent_size, self.latent_size),
+            nn.Linear(self.latent_size, self.num_classes),
+        )
+
+    def forward(self, test_input):
+        enc_output = self.embedding(test_input)
+        for enc_layer in self.encoders:
+            enc_output = enc_layer(enc_output)
+
+        class_token_embed = enc_output[:, 0]
+        return self.MLPHead(class_token_embed)
+
+
+class TrainEval:
+
+    def __init__(self, model, train_dataloader, val_dataloader, optimizer, criterion, epochs, device):
+        self.model = model
+        self.train_dataloader = train_dataloader
+        self.val_dataloader = val_dataloader
+        self.optimizer = optimizer
+        self.criterion = criterion
+        self.epoch = epochs
+        self.device = device
+
+    def train_fn(self, current_epoch):
+        self.model.train()
+        total_loss = 0.0
+        tk = tqdm(self.train_dataloader, desc="EPOCH" + "[TRAIN]" + str(current_epoch + 1) + "/" + str(self.epoch))
+
+        for t, data in enumerate(tk):
+            images, labels = data
+            images, labels = images.to(self.device), labels.to(self.device)
+            self.optimizer.zero_grad()
+            logits = self.model(images)
+            loss = self.criterion(logits, labels)
+            loss.backward()
+            self.optimizer.step()
+
+            total_loss += loss.item()
+            tk.set_postfix({"Loss": "%6f" % float(total_loss / (t + 1))})
+
+        return total_loss / len(self.train_dataloader)
+
+    def eval_fn(self, current_epoch):
+        self.model.eval()
+        total_loss = 0.0
+        tk = tqdm(self.val_dataloader, desc="EPOCH" + "[VALID]" + str(current_epoch + 1) + "/" + str(self.epoch))
+
+        for t, data in enumerate(tk):
+            images, labels = data
+            images, labels = images.to(self.device), labels.to(self.device)
+
+            logits = self.model(images)
+            loss = self.criterion(logits, labels)
+
+            total_loss += loss.item()
+            tk.set_postfix({"Loss": "%6f" % float(total_loss / (t + 1))})
+
+        return total_loss / len(self.val_dataloader)
+
+    def train(self):
+        best_valid_loss = np.inf
+        best_train_loss = np.inf
+        for i in range(self.epoch):
+            train_loss = self.train_fn(i)
+            val_loss = self.eval_fn(i)
+
+            if val_loss < best_valid_loss:
+                torch.save(self.model.state_dict(), "best-weights.pt")
+                print("Saved Best Weights")
+                best_valid_loss = val_loss
+                best_train_loss = train_loss
+        print(f"Training Loss : {best_train_loss}")
```

Comment:
could you add roughly the numbers you get for each of the losses if you run with default settings, it'll help debug any future regressions